gobblin-data-management/src/main/java/gobblin/data/management/trash/AsyncTrash.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.trash;
+
+import gobblin.util.Decorator;
+import gobblin.util.ExecutorsUtils;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.util.concurrent.MoreExecutors;
+
+
+/**
+ * Implementation of {@link Trash} that deletes files asynchronously and in parallel.
+ *
+ * <p>
+ *   This implementation is not built through {@link TrashFactory} because coder must be aware that the trash
+ *   implementation is asynchronous. However, internally it uses {@link TrashFactory} to instantiate the trash
+ *   implementation that will actually perform the deletes. This class acts as a {@link Decorator} of the
+ *   inner trash.
+ * </p>
+ *
+ * <p>
+ *   Trash methods will always return true, regardless of success of the actual trash operation. However, additional
+ *   methods are provided to get a future for the operation.
+ * </p>
+ */
+public class AsyncTrash implements GobblinProxiedTrash, Closeable, Decorator {
```

Comment:
The rate limiter should be implemented through the filesystem, not in the trash itself. I will change the thread pool executor.